### PR TITLE
Avoid hang during concurrent db_cancel_event()

### LINKED ISSUE #423: dbEvent: double cancel causes hang
A logic hole in dbEvent.c can combine with a race between concurrent `db_cancel_event()` and an event callback to cause `db_cancel_event()` to hang.  The following assertion will occasionally fail if a thread calls `db_cancel_event()` on two different subscriptions.  Not a double free, two separate free's of subscriptions which share the same dbEventCtx.

While I am not sure I can completely explain the sequence of events, I think this issue arises from an attempt to handle self-cancel and concurrent cancel with the same logic.  One fix I will explore is to move self-cancel detection from the common `struct event_user` to the individual `struct evSubscrip`.

```diff
diff --git a/modules/database/src/ioc/db/dbEvent.c b/modules/database/src/ioc/db/dbEvent.c
index 6dab45104..7cbbda91b 100644
--- a/modules/database/src/ioc/db/dbEvent.c
+++ b/modules/database/src/ioc/db/dbEvent.c
@@ -620,6 +620,7 @@ void db_cancel_event (dbEventSubscription event)
     assert ( pevent->npend == 0u );
 
     if ( pevent->ev_que->evUser->taskid == epicsThreadGetIdSelf() ) {
+        assert(!pevent->ev_que->evUser->pSuicideEvent);
         pevent->ev_que->evUser->pSuicideEvent = pevent;
     }
     else {
```

```
A call to 'assert(!pevent->ev_que->evUser->pSuicideEvent)'
    by thread 'qsrvSingle' failed in ../db/dbEvent.c line 623.
Dumping a stack trace of thread 'qsrvSingle':
[    0x7f5dcc2a352a]: /home/dev/epics/pvxs/test/O.linux-x86_64-debug/../../../base-git/lib/linux-x86_64-debug/libCom.so.3.22.1(epicsBackTrace+0x20)
[    0x7f5dcc2a33ac]: /home/dev/epics/pvxs/test/O.linux-x86_64-debug/../../../base-git/lib/linux-x86_64-debug/libCom.so.3.22.1(epicsStackTrace+0x6f)
[    0x7f5dcc299d21]: /home/dev/epics/pvxs/test/O.linux-x86_64-debug/../../../base-git/lib/linux-x86_64-debug/libCom.so.3.22.1(epicsAssert+0x5e)
[    0x7f5dcc31dc6b]: /home/dev/epics/pvxs/test/O.linux-x86_64-debug/../../../base-git/lib/linux-x86_64-debug/libdbCore.so.3.22.1(db_cancel_event+0x1fb)
[    0x7f5dcc6882a1]: /home/dev/epics/pvxs/test/O.linux-x86_64-debug/../../lib/linux-x86_64-debug/libpvxsIoc.so.1.2(_ZZN4pvxs3ioc12Subscription9subscribeEPvRKNS0_7ChannelEPFvS2_P9dbChanneliP12db_field_logES2_jENUlS2_E_clES2_+0x2b)
```

### PR DESCRIPTION
**Pull request by mdavidsaver**
Intended to fix #423.  A further correction to #424 (fab8fd7102143a661a81cf7163dd889601eb015c).

My attempt to handle concurrent `db_cancel_event()` just didn't work.  I had been hoping to avoid having to allocate (temporarily) an `epicsEventId` for each waiter.  However, I can't see a way around this.  So the single `pflush_sem` is replaced with a list of temporary events.

When a concurrent cancel is detected, the caller will (hopefully) allocate an epicsEventId and add this to the list of waiters.  It will then poke the worker in case it has already gone idle, and then wait for the corresponding notification that the worker is going idle again.

### CODE REVIEW

File: modules/database/src/ioc/db/dbEvent.c [hunk 1]
```diff
@@ -592,23 +592,32 @@ void db_cancel_event (dbEventSubscription event)
...
-             * So uncondtionally re-trigger.  This will result in one spurious
-             * wakeup for each cancellation.
-             */
-            epicsEventTrigger(evUser->pflush_sem);
+            epicsEventMustTrigger(evUser->ppendsem);
+
+            if(wait.wake) {
+                epicsEventMustWait(wait.wake);
+            } else {
+                epicsThreadSleep(0.01); /* ick. but better than cantProceed() */
```
Comment (mdavidsaver) [inline modules/database/src/ioc/db/dbEvent.c:613]: I know this looks odd.  I didn't want to call `epicsEventMustCreate()`, and chance a `cantProceed()`.  This seemed like the least bad option.

File: modules/database/src/ioc/db/dbEvent.c [hunk 2]
```diff
@@ -1041,11 +1050,18 @@ static void event_task (void *pParm)
         pendexit = evUser->pendexit;
 
         evUser->pflush_seq++;
+        if(ellCount(&evUser->waiters)) {
+            /* hold lock throughout to avoid race between event trigger and destroy */
+            ELLNODE *cur;
+            for(cur = ellFirst(&evUser->waiters); cur; cur = ellNext(cur)) {
+                event_waiter *w = CONTAINER(cur, event_waiter, node);
+                if(w)
+                    epicsEventMustTrigger(w->wake);
```
Comment (anjohnson) [inline modules/database/src/ioc/db/dbEvent.c]: In the body of the `for (…; cur; …)` loop we know that `cur` can't be NULL, so how could `w` (which I think should point to the same address as `cur`) ever be NULL? Shouldn't that `if(w)` be `if (w->wake)` instead?

Comment (mdavidsaver) [reply modules/database/src/ioc/db/dbEvent.c]: You are correct, it should.

Comment (mdavidsaver) [reply modules/database/src/ioc/db/dbEvent.c]: Fixed.

**Comment by anjohnson** [comment]
Core Group: Approved.

**Pull request merged**